### DISCUSSION TITLE
Collect requirements/idea/features for v2.9.0

### DISCUSSION THREAD

**Discussion by OrlinVasilev** [goharbor/harbor#18396] [Ideas]
Hello Community,

as we are almost there with release v2.8 and we have plan and roadmap for v2.9 it's time to brainstorm the ideas and requirements for the feature list for 2.9!

Please take a look for all candidates for 2.8 that will be transferred to 2.9:
[goharbor/harbor/issues (is:open sort:updated-desc label:candidate/2.8.0)](https://github.com/goharbor/harbor/issues?q=is%3Aopen+sort%3Aupdated-desc+label%3Acandidate%2F2.8.0+)
And 2.9 candidates already added:
[goharbor/harbor/issues (is:open sort:updated-desc label:candidate/2.9.0)](https://github.com/goharbor/harbor/issues?q=is%3Aopen+sort%3Aupdated-desc+label%3Acandidate%2F2.9.0+)

Please share your thoughts here so we can all work on proper roadmap and achieving it!
Keep in mind that Harbor is community developed and driven project so not everything can be implemented right away :)
If you have an idea or need and eager to work on, we will support you and help you get there!

Thank you being essential part of the community!

Orlix
Harbor Community Manager

**Comment by Vad1mo** [comment]
### Immutability and Deletion with More Capable Retention Policies.

The purpose of Immutable tags is to prevent images from being deleted. 
For this reason, we declined proposals like https://github.com/goharbor/harbor/issues/18123 in the past. 

I would like to make a proposal that, on one hand, keeps the designed promise of not deleting an immutable image, while on the other hand, allows cleaning up images.


### Feature Request

Extend the retention Policies capabilities to be applied to immutable tags and labels.

This would feature would make it possible to create a retention policy that will remove immutability or any other label. The free tag can then be processed further, and other immutability rules can be applied.  
One variation of the request would be to be able to create retention polices that ignore immutability tags, but in that wouldn't as flexible as retention policies that can be applied on immutability tag or labels.

**Comment by Vad1mo** [comment]
### Popular Artifact Types like Helm Charts as Tabs in Harbor 

It was very convenient and user-friendly to be able to see Helm Charts in a separate tab.

From the UX/DX perspective and considering the bigger role of OCI artifacts in the Cloud Native ecosystem. 
We should extend Harbor to be able to display different types of OCI artifact as dedicated tabs.

OCI registries store nowadays different types of artifacts apart from container images, eg. ORAS, Helm OCI, Singularity, ML Data Models and so on. 

Currently, Harbor has only tab containing all the different artifacts types in one single list.

<img width="1853" alt="Bildschirmfoto 2022-05-28 um 12 15 48" src="https://user-images.githubusercontent.com/1492007/170821152-13146560-4065-4c18-9d72-041e924d79d9.png">

Having a dropdown to filter by artifact type is, in my opinion, not a solution to how user interact with Harbor. 

My proposal for Harbor 2.9 is that we should extend Harbor in a way that a user (project admin) can configure per-project what tabs should be displayed, beyond the default one. (it can also be done automatically) 

<img width="1773" alt="Bildschirmfoto 2022-05-28 um 12 18 48" src="https://user-images.githubusercontent.com/1492007/170821142-a074a0dd-1e69-4687-bc00-b4ea377fe8ae.png">


The results would be that an organization with ML teams could config their project to only see ML Data for TensorFlow and the software engineering team would be fine with only seeing container images and WebAssembly Artifacts, whereas the Ops teams would see Helm Charts, OPA policies.

**Comment by lengrongfu** [comment]
Is it possible to set a read-only storage ratio for the system? When the used storage reaches this value, the image cannot be pushed, but can only be pulled, because the system will crash when the disk space is used up, and it is not easy to recover.

**Comment by lengrongfu** [comment]
Implement image upload/download on the Harbor UI interface.
https://github.com/goharbor/harbor/issues/17028
> I recently implemented this feature with the help of `skopeo`, if the community needs this feature, then I can add this feature.

**Reply by OrlinVasilev** [reply]
you can open a proposal in the [community](https://github.com/goharbor/community/) repository and work on this one :) 
check the process here : https://github.com/goharbor/harbor/blob/main/CONTRIBUTING.md#develop-and-propose-new-features

**Reply by lengrongfu** [reply]
Thank you for your suggestion, I will follow the process.

**Reply by OrlinVasilev** [reply]
@lengrongfu fantastic :))) thank  you so much!

**Comment by lengrongfu** [comment]
I think this function is meaningful, can we implement it?
My idea is that we can add a function in the system settings, which can set the configuration of the total available storage space, and set the total usage to a percentage value to enter the read-only state and reject all push actions.
https://github.com/goharbor/harbor/issues/16671

**Reply by OrlinVasilev** [reply]
@lengrongfu can you work on this one ?

**Reply by lengrongfu** [reply]
I can. I would write a proposal first, and then develop it after approval.

**Reply by OrlinVasilev** [reply]
nice sounds great :)) Thank you @lengrongfu 
@KaulAnmol @msha01 for visibility and planning :)

**Comment by IwanSE** [comment]
Hi all!

@OrlinVasilev, @Vad1mo

Is it possible to rework cache projects - so that 3-6 or more caches can be created in one project?
This will reduce the number of cache projects to one - which will be a big plus for the harbor
Example for creation from UI.

I'd love to hear feedback!

![изображение](https://github.com/goharbor/harbor/assets/18316714/981f788b-e944-40b1-a10b-53d937160b12)

**Reply by OrlinVasilev** [reply]
that's fantastic idea :)) not sure how complicated it will be to be implemented but it's GREAT!

**Reply by OrlinVasilev** [reply]
@Vad1mo @wy65701436 @chlins @msha01 - if we can make it in will be fantastic :)

**Comment by flbla** [comment]
## Robot Accounts replication between Harbor instances
I am interested in replicating multiple Harbor instances (with one serving as the primary read/write and others as read-only). 
This replication process will include the automatic replication of projects along with their associated artifacts using the actual replication system.

### Feature Request: Robot accounts replication

In addition to project replication, I would like to have a replication system that can handle robot accounts (system and project).

This sync process could in the future handle more feature (like users, groups, ...)
